compiler/rustc_typeck/src/check/op.rs

```diff
@@ -435,39 +434,35 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         // concatenation (e.g., "Hello " + "World!"). This means
                         // we don't want the note in the else clause to be emitted
                     } else if let [ty] = &visitor.0[..] {
-                        if let ty::Param(p) = *ty.kind() {
-                            // Check if the method would be found if the type param wasn't
-                            // involved. If so, it means that adding a trait bound to the param is
-                            // enough. Otherwise we do not give the suggestion.
-                            let mut eraser = TypeParamEraser(self, expr.span);
-                            let needs_bound = self
-                                .lookup_op_method(
-                                    eraser.fold_ty(lhs_ty),
-                                    &[eraser.fold_ty(rhs_ty)],
-                                    Op::Binary(op, is_assign),
-                                )
-                                .is_ok();
-                            if needs_bound {
-                                suggest_constraining_param(
-                                    self.tcx,
-                                    self.body_id,
+                        // Look for a TraitPredicate in the Fulfillment errors,
+                        // and use it to generate a suggestion.
+                        //
+                        // Note that lookup_op_method must be called again but
+                        // with a specific rhs_ty instead of a placeholder so
+                        // the resulting predicate generates a more specific
+                        // suggestion for the user.
+                        let errors = self
+                            .lookup_op_method(lhs_ty, &[rhs_ty], Op::Binary(op, is_assign))
+                            .unwrap_err();
+                        let predicates = errors
+                            .into_iter()
+                            .filter_map(|error| error.obligation.predicate.to_opt_poly_trait_pred())
+                            .collect::<Vec<_>>();
+                        if !predicates.is_empty() {
+                            for pred in predicates {
+                                self.infcx.suggest_restricting_param_bound(
```

Comment:
I think we should teach this function about associated type bounds. You could filter all projection predicates for one that is for the current trait and type and if there is such a projection predicate, extend the suggestion with the assoc bound

Reply:
One issue I'm encountering. I don't think there's a sound way to determine what the appropriate `Output` type should be. The function `check_overloaded_binop` attempts to do this, but it can only compute `return_ty` if the initial `lookup_op_method` returns a `MethodCallee`, which it does not in the case being addressed. 

Is there some possibility here that I'm missing?

Reply:
I'm hoping that the correct associated type bound is in the `errors` list. Try dumping the entire list and checking if it contains `PredicateKind::Projection` that match the `pred` in the loop here.

Reply:
Yes I've checked, and a `Projection` predicate does not already exist. This is because the relevant predicates are generated by the function [`obligation_for_method`](https://github.com/rust-lang/rust/blob/master/compiler/rustc_typeck/src/check/method/mod.rs#L298). That function would have to be extended to produce a `Projection` predicate, except it can only do so if it knows about the expected associated types of a method, which it cannot in this case.

Reply:
I think I made some related changes in https://github.com/rust-lang/rust/pull/90006/files, which won't be landing until I manage to get rid of the perf regression.